File: src/drowning.cpp
#include <Rcpp.h>
include "ndGrid.cpp";
using namespace Rcpp;
using namespace std;


double floodingLevel (NumericVector coords, const IntegerVector flooded, const NumericMatrix waterLevel, NumericVector gridStep,IntegerVector d_){
	int position;
	double floodLevel;
	position = coord2idx(coords,gridStep, d_);
	if (is_true(any(position == flooded))){

		//how are supposed to be stored the waterlevels from previous run?

		//TRIVIAL!!!!! WARNING!!!!
		floodLevel=1;
	}


	return floodLevel;

}

//prepare the grid of values




//find the minimum
void valley(NumericVector meanParam, Function floodingLevel, NumericVector lowerParam, NumericVector upperParam,NumericVector gridStep,IntegerVector d_,IntegerVector & flooded ){
	NumericVector giudecca;
	List giudeccaL;
	giudeccaL=abcNVm(meanParam,
	floodingLevel,
	    lowerParam,
	    upperParam,
	    //int FoodNumber   =
	    20,
	    //int limit        =
	    100,
	    //int maxCycle     =
	    1000,
	    //int criter       =
	    50,
	    //NumericVector
	    gridStep,
	    //NumericVector
	    d_ // , double tol=1e-10
	);
	giudecca=as<NumericVector>(giudeccaL["globalParam"]);
	//include minimum in the grid

	IntegerVector giudIdx;
	giudIdx=coord2idx(giudecca,gridStep,d_);
	//flooding the bottom
	IntegerVector flooding=IntegerVector::create();
	NumericVector groundLevel=NumericVector::create();

	flooding.insert(giudIdx);
	groundLevel.begin(as<double>(giudeccaL["value"]));

	waterShot(flooded,flooding,gridStep,d_,floodingLevel);

	while(flooding[0]>flooding[1])
		waterShot(flooded,flooding,gridStep,d_,floodingLevel);
	flooding.erase(0);
	groundLevel.erase(0);
	double hyperLiter=0;
	for(int i=1; i<groundLevel.size(); i++)
		hyperLiter += (groundLevel[0]-groundLevel[i]);
	//print(hyperLiter);

}

//starting loop
NumericVector waterShot(const IntegerVector flooded , IntegerVector & flooding, NumericVector gridStep, IntegerVector d_, Function model ){
	//look for neighs
	NumericVector neigHeight;
	NumericVector neighs;
	neighs=getNeighbours(flooded[0], d_);
	//check neighs for flooded
	for (int i =(neighs.size()-1);i==0;i--){
		if(is_true(any(neighs[i]==flooding))){
			neighs.erase(i);
		} else {
			if(is_true(any(neighs[i]==flooded)){
				neigHeight.pushback(model(idx2coord(neighs[i],d_,gridStep)));
			}
	}
	double candidate= neigHeight[0];
	int lowNeigh = 0;

	for (int i=1; i<neigHeight;i++){
		if(neigHeight[i]<candidate)
			lowNeigh =i;
	}
	//flood lowNeigh

	flooding.insert(lowNeigh);
	groundLevel.insert(candidate);

}




//minimum of neighs

//waterfill

//if valley filled, mmove to

File: R/drowning.R
cppFunction('
	NumericVector ritaglia(NumericVector nomi, NumericVector obiettivi){
		NumericVector listarella;
		for (int i{0}; i<= nomi.size(); i++){
			for ( int j{0}; j<=obiettivi.size();j++){
				if (nomi[i] == obiettivi[j]){
					listarella.push_back(i+1);
				}
			}
		}
		return listarella;
	}
')




gridABCD<-60
meanParam<-colMeans(desiKg)
d_<-meanParam
gridStep<-meanParam
for(parameter in 1:length(meanParam)){
	dimensioneDimensione<-ceiling((gridABCD/rowSums(sensiB[1,]))*sensiB[,parameter])+1
	if(parameter >=2){
		d_[parameter]<-dimensioneDimensione*d_[parameter-1]
	}else{
		d_[parameter]<-dimensioneDimensione
	}

	gridStep[parameter]<-(max(desiKg[,parameter])-min(desiKg[,parameter]))/dimensioneDimensione
}
d_<-as.integer(d_)


#here qe are going to calculate the valley extension...
valley<-function(meanParams,morphLevel,lowerParam,upperParam,gridStep,d_,flooded){
		floodingLevel<-function(idx){
			if(length(flooded[,1])>=2){
				water<-NULL
				if(any(idx==unlist(flooded[,3]))){
					for(vallata in length(flooded[,1])){
						if(any(unlist(idx==flooded[vallata,3]))){
							water<-max(c(water,flooded[vallata,2]))
						}
					}
				}else{
					water<-morphLevel(idx2coord(idx,d_,gridStep))
				}
			}else{
				water<-morphLevel(idx2coord(idx,d_,gridStep))
			}
			print(water)
			return(water)
		}

	#find the very bottom  point
	giudecca<-abcNVm(par=meanParams,fn=floodingLevel,lb=lowerParam,ub=upperParam)
	giudIdx<-as.integer(coord2idx(giudecca,gridStep,d_))
	flooding<-giudIdx
	waterLevel<-floodingLevel(flooding[1])

	#inizializing the chain by flooding giudecca
	surface<-waterLevel
	flow<-NULL
	candidate<-NULL
	candHei<-NULL




	while(waterLevel <= surface){
		waterLevel<-surface
		flooding<-as.integer(c(flow,flooding))
		neighs<-as.integer(getNeighbours(flooding[1],d_))
		#neighs<-neighs[-c(ritaglia(neighs,flooding))]
		neiGround<-neighs
		for (neG in 1:length(neiGround)){
			neiGround[neG]<-floodingLevel(neiGround[neG])
		}
		candidate<-c(candidate,neighs)
		candHei<-c(candHei,neiGround)
		flow<-candidate[which(candHei==min(candHei))]
		surface<-min(candHei)
	#drop in the whole valley if a valley is encountered.... (optimization available in near future!)
	}
	output<-data.frame(fondo=floodingLevel(giudIdx),quota=waterLevel,volume=list(flooding))
	return(output)
}


orografia<-data.frame(fondo=c(0),quota=c(0),volume=c(0))
for(pippo in 1:2){
	print(pippo)
	orografia<-rbind(orografia,
		valley(startHyp,krfunBra,minHyp,maxHyp,gridStep,d_,orografia))
}
